src/ZkFold/Cardano/UPLC/Wallet.hs

```diff
@@ -4,37 +4,78 @@
 
 module ZkFold.Cardano.UPLC.Wallet where
 
-import           GHC.Generics                     (Generic)
+import           GHC.Generics                          (Generic)
+import           PlutusLedgerApi.V1.Value              (lovelaceValue)
 import           PlutusLedgerApi.V3
-import           PlutusTx                         (makeIsDataIndexed, makeLift)
-import           PlutusTx.Prelude                 hiding (toList, (*), (+))
-import           Prelude                          (Show, undefined)
+import           PlutusTx                              (makeIsDataIndexed, makeLift)
+import qualified PlutusTx.AssocMap                     as M
+import           PlutusTx.Prelude                      hiding (toList, (*), (+))
+import           Prelude                               (Show)
 
-import           ZkFold.Cardano.OnChain.BLS12_381 (F)
+import           ZkFold.Cardano.UPLC.ForwardingScripts (forwardingReward)
 
 data WalletSetup = WalletSetup
   { wsPubKeyHash :: BuiltinByteString
-  , wsWeb2UserId :: F
+  , wsWeb2UserId :: BuiltinByteString
   } deriving stock (Show, Generic)
 
 makeLift ''WalletSetup
 makeIsDataIndexed ''WalletSetup [('WalletSetup,0)]
 
-data WalletRedeemer = SpendWithSignature BuiltinByteString | SpendWithWeb2Token F
+data SpendingCreds = SpendWithSignature BuiltinByteString | SpendWithWeb2Token BuiltinByteString
   deriving stock (Show, Generic)
 
-makeIsDataIndexed ''WalletRedeemer [('SpendWithSignature,0),('SpendWithWeb2Token,1)]
+makeLift ''SpendingCreds
+makeIsDataIndexed ''SpendingCreds [('SpendWithSignature,0),('SpendWithWeb2Token,1)]
+
+data WalletRedeemer = WalletRedeemer
+    { wrTxDate :: BuiltinByteString
+    , wrCreds  :: SpendingCreds
+    }
+  deriving stock (Show, Generic)
+
+makeLift ''WalletRedeemer
+makeIsDataIndexed ''WalletRedeemer [('WalletRedeemer, 0)]
+
+type Web2Check =
+       BuiltinByteString -- ^ token
+    -> BuiltinByteString -- ^ date (64 bits)
+    -> BuiltinByteString -- ^ client id
+    -> Bool
 
--- TODO: Implement the smart wallet script.
 {-# INLINABLE wallet #-}
-wallet :: WalletSetup -> WalletRedeemer -> ScriptContext -> Bool
-wallet = undefined
+wallet :: Web2Check -> WalletSetup -> WalletRedeemer -> ScriptContext -> Bool
+wallet zkpCheck WalletSetup{..} WalletRedeemer{..} ctx@(ScriptContext TxInfo{..} _ scriptInfo) =
+    signedCorrectly && valueIsCorrect && fwd
+    where
+        signedCorrectly =
+            case wrCreds of
+              SpendWithSignature sign  -> verifyEd25519Signature wsPubKeyHash (getTxId $ txInfoId) sign
+              SpendWithWeb2Token token -> zkpCheck token wrTxDate wsWeb2UserId
```

Comment:
I am guessing, the `zkpCheck` function should accept a proof as one of the arguments (instead of `token`?), because all of its arguments are public and recorded on the blockchain. The private user info must not be included.